### ISSUE TITLE
op_ref false positive with deref coercions when comparing

### ISSUE THREAD

**Issue by euclio** [rust-lang/rust-clippy#2597]
Perhaps related to #2042, but it seemed different enough to open a new issue.

With the following code: https://play.rust-lang.org/?gist=35d367db95cd56394d0bb9a05247b94c&version=nightly

```rust
fn main() {
    let a: &str = "abc";
    let b: String = "abc".to_owned();
    println!("{}", a > &b);
}
```

Clippy complains that the reference should be removed, but removing the reference causes the code to stop compiling.

**Comment by oli-obk** [comment]
Clippy on playpen is down. If memory serves right, isn't it also suggesting to deref the other side?

**Comment by euclio** [comment]
The error message reported by clippy is:

```
┌ euclio@apollo ~/repos/op_ref_test (git:master)
└╌╌┄┄ ❯❯ cargo +nightly clippy
   Compiling op_ref_test v0.1.0 (file:///Users/anrussell/repos/op_ref_test)
warning: taken reference of right operand
 --> src/main.rs:6:20
  |
6 |     println!("{}", a > &b);
  |                    ^^^^--
  |                        |
  |                        help: use the right value directly: `b`
  |
  = note: #[warn(op_ref)] on by default
  = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.191/index.html#op_ref

    Finished dev [unoptimized + debuginfo] target(s) in 1.37 secs
```

**Comment by oli-obk** [comment]
We need to suggest a deref on the lhs here, too.

**Comment by oli-obk** [comment]
Dereferencing both sides works. We can probably figure this out by checking for autoref coercions

**Comment by mikerite** [comment]
I didn't know that was possible but what benefit does `*a > *b` have over `a > &b`?
Also this seems to be a regression of #1665.

**Comment by oli-obk** [comment]
The advantage is one less temporary variable (the comparison operator internally takes a reference to its arguments).

**Comment by jsgf** [comment]
I just ran into this with https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=890234c7ce6d7ccde1d036ae41d782d4

```
pub fn false_clippy_op_ref<T: Ord + PartialOrd>(array: &[T], val: &T, idx: usize) -> bool {
    &array[idx] < val
}
```

**Comment by Luro02** [comment]
I think I ran into this too
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d57a5e67a0a9e7d50e4bf85a76c7602a

```rust
pub fn example_problematic_lint(value: &String, rhs: &str) -> bool {
    &**value < rhs
}


fn main() {
    dbg!(example_problematic_lint(&"Hello".to_string(), "Worl"));
}
```

Clippy output
```
   Checking playground v0.0.1 (/playground)
warning: needlessly taken reference of left operand
 --> src/main.rs:3:5
  |
3 |     &**value < rhs
  |     --------^^^^^^
  |     |
  |     help: use the left value directly: `**value`
  |
  = note: `#[warn(clippy::op_ref)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#op_ref

    Finished dev [unoptimized + debuginfo] target(s) in 0.57s
```

**Comment by flip1995** [comment]
This is probably fixed by #4878, but needs more tests, e.g. https://github.com/rust-lang/rust-clippy/issues/2597#issuecomment-519273069.

An improvement of this lint would still be to suggest a deref on both sides, when possible.

**Comment by ssokolow** [comment]
I just ran into this with a newtype I defined:

```
warning: taken reference of right operand
   --> src/lib.rs:208:37
    |
208 |         || f64::from(stats.words) < f64::from(stats.urls) * &*config.min_words_per_url
    |                                     ^^^^^^^^^^^^^^^^^^^^^^^^--------------------------
    |                                                             |
    |                                                             help: use the right value directly: `*config.min_words_per_url`
    |
    = note: `#[warn(clippy::op_ref)]` implied by `#[warn(clippy::all)]`
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#op_ref
```

```
error[E0277]: cannot multiply `f64` by `config::newtypes::PositiveF64`
   --> src/lib.rs:208:59
    |
208 |         || f64::from(stats.words) < f64::from(stats.urls) * config.min_words_per_url
    |                                                           ^ no implementation for `f64 * config::newtypes::PositiveF64`
    |
    = help: the trait `std::ops::Mul<config::newtypes::PositiveF64>` is not implemented for `f64`
    = help: the following other types implement trait `std::ops::Mul<Rhs>`:
              <&'a f32 as std::ops::Mul<f32>>
              <&'a f64 as std::ops::Mul<f64>>
              <&'a i128 as std::ops::Mul<i128>>
              <&'a i16 as std::ops::Mul<i16>>
              <&'a i32 as std::ops::Mul<i32>>
              <&'a i64 as std::ops::Mul<i64>>
              <&'a i8 as std::ops::Mul<i8>>
              <&'a isize as std::ops::Mul<isize>>
            and 49 others
```

Granted, I'll implement `Mul` after posting this, but I'm sure there are situations where that's not an option.

**Comment by Jarcho** [comment]
Closing since this was fixed with #4878.

@ssokolow The suggestion was to use `*config.min_words_per_url`, not `config.min_words_per_url`.
